### ISSUE TITLE
Provide an ocean-d2 repo

### ISSUE THREAD

**Issue by hatem-oraby-sociomantic** [sociomantic-tsunami/ocean#84]
Currently if you have a D2 only application and you decided to use ocean then you have to do the following:
- Make sure that you have [d1to2fix](https://github.com/sociomantic-tsunami/d1to2fix/) installed on your system
- Add Ocean as a submodule
- Go to the ocean submodule directory and checkout the desired branch
- Add [Makd](https://github.com/sociomantic-tsunami/makd/) as a submodule to ocean
- Run `make d2conv`
- If your project is under version-control, then you need to commit the converted code so you can have none-dirty commits.
- And only at that point you are ready to use ocean.

The above procedure may deter a lot of D2 users who would like to test Ocean. It would be much more convenient to have a new ocean-d2 repo that contains converted snapshots of the released tags of this repo.
Syncing the d2 repo can be automatically maintained by Travis.
Also, I don't know if current ocean is in dub, but certainly the converted Ocean can live there.

fyi @leandro-lucarella-sociomantic 

---
**Disclaimer:** I'm a direct beneficiary of this issue. I will give a (2x-slot-)lightning talk about Ocean in DConf. I'd like the ease of using Ocean to be one of the selling points.

**Comment by leandro-lucarella-sociomantic** [comment]
Yeah, I think simplifying access to D2 users is key for people to even dare to have an eye on Ocean, so I think a plan like this could be a very good start. We still have the major issue of not being able to use upstream compilers yet, but hopefully that will be sorted out not too far in the future if the DIP about `@future` gets accepted and implemented soon so we can solve the `Throwable.message()` issue.

**Comment by mihails-strasuns-sociomantic** [comment]
Last time it was discussed, automatic generation of branches with `-d2` suffix in the same repo was the intended approach. I think it works better than dedicated repo because it prevents confusion where to create issues/PRs and makes it more obvious that content is automatically generated.

**Comment by gavin-norman-sociomantic** [comment]
>I will give a (2x-slot-)lightning talk about Ocean in DConf. I'd like the ease of using Ocean to be one of the selling points.

Oh that's great!

**Comment by hatem-oraby-sociomantic** [comment]
> Last time it was discussed, automatic generation of branches with -d2 suffix in the same repo was the intended approach. I think it works better than dedicated repo because it prevents confusion where to create issues/PRs and makes it more obvious that content is automatically generated.

Both would work fine, however there are few pros and cons to each side:
- Pros for a separate D2 repo:
  - Creating a separate repo would remove noise from the original repo
  - I think it'll be easier to add it to dub since it will also have correct version number (without a `-d2` suffix).
  - We can disable the issues and PRs in the new repo and we can make Travis CI inject a line to the readme file saying that it's a view-only repo and links to the original repo.

- Pros for same repo:
  - D1 applications that are converted to D2 can reference the commits of the D2 branches. They can have a clean version-info as a result.

**Comment by nemanja-boric-sociomantic** [comment]
> D1 applications that are converted to D2 can reference the commits of the D2 branches. They can have a clean version-info as a result.

This only apply to applications that are not converted to D2 during build time, but have D2 commited code, though, and at that point you could reference the another repo anyway.

**Comment by leandro-lucarella-sociomantic** [comment]
We tried in the past with putting converted D2 code in the same repo but for me it didn't worked well, it pollutes a lot the blessed repo with what is basically garbage for developers. These automatically converted branches would only be useful for pure-users.

To keep things simple I wonder if just adding a `.tag.gz` with the converted code to the GitHub releases wouldn't be enough. We could do this automatically via travis too (I think). Worse case as a daily cron job (I think travis support cron job).

**Comment by hatem-oraby-sociomantic** [comment]
> To keep things simple I wonder if just adding a .tag.gz with the converted code to the GitHub releases wouldn't be enough

I think this would be inconvenient. You can't download it through git cli and you have to extract and set it up. Also you we won't be able to add it to dub (unless dub has support for `tar.gz` that I'm not aware or of).

> We could do this automatically via travis too (I think). Worse case as a daily cron job (I think travis support cron job).

It claims to support it: https://docs.travis-ci.com/user/cron-jobs/

**Comment by leandro-lucarella-sociomantic** [comment]
Can't you upload stuff to dub?

**Comment by hatem-oraby-sociomantic** [comment]
Not that I'm aware of: https://code.dlang.org/publish

**Comment by mihails-strasuns-sociomantic** [comment]
tarball also doesn't fit well with the idea that D2-only applications will eventually depend on already converted version in a submodule

**Comment by hatem-oraby-sociomantic** [comment]
I tried to implement a script that takes each tag and converts to D2 it and sync it to a remote repo:
https://github.com/oraby/ocean/commit/fe7219f3a01d8beb2417091dc8ce4c02078d3bb3

@leandro-lucarella-sociomantic  (and anyone else who maintains the CI): if we agreed to use a remote repo, does that script conver everything needed to make a cron-job or there are more missing steps?

**Comment by leandro-lucarella-sociomantic** [comment]
I'll have a look.

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, I created the repo already, in case someone else can review/test before I can: https://github.com/sociomantic-tsunami/ocean-d2

**Comment by hatem-oraby-sociomantic** [comment]
> BTW, I created the repo already, in case someone else can review/test before I can: https://github.com/sociomantic-tsunami/ocean-d2

How can I test it? or is it only for those with write permissions?

Also, can you disable Issues and Wiki on that repo.

**Comment by leandro-lucarella-sociomantic** [comment]
Yes!

**Comment by leandro-lucarella-sociomantic** [comment]
You should probably add a [pull request template](https://github.com/blog/2111-issue-and-pull-request-templates) too to discourage people from opening a PR.

**Comment by hatem-oraby-sociomantic** [comment]
> You should probably add a pull request template too to discourage people from opening a PR.

This sounds like a new idea as well, I can write it but I need someone with write-access to commit it.

**Comment by llucax** [comment]
Depending on how do you plan to integrate new code in the new ocean-d2 repo, you can do it yourself via the cron-job.

**Comment by llucax** [comment]
I just checked in Travis, and the cron job just means running the same `.travis.yml` but at a regular interval (and it must be enabled in a per-branch basis, so each time a new branch is added we need to enable the cron job for the new branch), so the whole sync procedure doesn't seem to be a good fit for this.

I think the way to go is to just see if what it's currently being built is a tag and in that case do the conversion and push to the ocean-d2 repo. Maybe we can even just mirror this repo and for each tag just add an extra commit in the ocean-d2 repo with the conversion and name the tag `<origina-tag>-d2` or something.

**Comment by llucax** [comment]
Checking for `$TRAVIS_TAG` should do the trick (https://docs.travis-ci.com/user/environment-variables#Default-Environment-Variables).

**Comment by oraby** [comment]
> I just checked in Travis, and the cron job just means running the same .travis.yml but at a regular interval (and it must be enabled in a per-branch basis, so each time a new branch is added we need to enable the cron job for the new branch), so the whole sync procedure doesn't seem to be a good fit for this

bummer, I've just saw that now after I've finished the sync script.
Here it is anyway:
https://github.com/oraby/ocean/commits/cli_sync

> I think the way to go is to just see if what it's currently being built is a tag and in that case do the conversion and push to the ocean-d2 repo

what I don't understand is when will travis build for a tag? I thought that branches are only tested and then tags are added later through direct pushing to the repo.

> Maybe we can even just mirror this repo and for each tag just add an extra commit in the ocean-d2 repo with the conversion and name the tag <origina-tag>-d2 or something.

As far as I understand, this will break dub as dub looks for SemVer tags.

**Comment by llucax** [comment]
> bummer, I've just saw that now after I've finished the sync script.

😢 

> Here it is anyway:
> https://github.com/oraby/ocean/commits/cli_sync

I made some comments anyway too. The good thing about this approach is, if we actually put the scripts in the ocean-d2 repo, is we keep this repo clean from this. But given that it is a bit more complex that way, I'm not sure it's justified.

> what I don't understand is when will travis build for a tag? I thought that branches are only tested and then tags are added later through direct pushing to the repo.

Exactly when you push the tag, it is considered a branch by travis, see https://travis-ci.org/sociomantic-tsunami/ocean/branches for example.

> As far as I understand, this will break dub as dub looks for SemVer tags.

That's still semver, although I used the wrong separator (`-` is for [pre-releases](http://semver.org/#spec-item-9)). I suggest using `+d2.auto` (`+` is for [build metadata](http://semver.org/#spec-item-10), since in theory the version is the same, it might make sense).
